packages/gatsby/src/redux/actions/restricted.js

```diff
@@ -0,0 +1,269 @@
+ *           }
+ *         },
+ *         author: {
+ *           type: 'AuthorJson'
+ *           extensions: {
+ *             link: {},
+ *           },
+ *         }
+ *         date: {
+ *           type: 'Date!'
+ *           extensions: {
+ *             dateformat: {},
+ *           },
+ *         },
+ *         published: 'Boolean!',
+ *         tags: '[String!]!',
+ *       }
+ *     }),
+ *     schema.buildObjectType({
+ *       name: 'AuthorJson',
+ *       fields: {
+ *         name: 'String!'
+ *         birthday: {
+ *           type: 'Date!'
+ *           extensions: {
+ *             dateformat: {
+ *               locale: 'ru',
+ *             },
+ *           },
+ *         },
+ *       },
+ *       interfaces: ['Node'],
+ *       extensions: {
+ *         infer: false,
+ *       },
+ *     }),
+ *   ]
+ *   createTypes(typeDefs)
+ * }
+ */
+actions.createTypes = (
+  types:
+    | string
+    | GraphQLOutputType
+    | GatsbyGraphQLType
+    | Array<string | GraphQLOutputType | GatsbyGraphQLType>,
+  plugin: Plugin,
+  traceId?: string
+) => {
+  return {
+    type: `CREATE_TYPES`,
+    plugin,
+    traceId,
+    payload: types,
+  }
+}
+
+const withDeprecationWarning = (action, api, allowedIn) => (...args) => {
+  report.warn(
+    `Calling \`${action}\` in the \`${api}\` API is deprecated. ` +
+      `Please use: ${allowedIn.map(a => `\`${a}\``).join(`, `)}.`
+  )
+  return actions[action](...args)
+}
+
+const withErrorMessage = (action, api, allowedIn) => () =>
+  // return a thunk that does not dispatch anything
+  () => {
+    report.error(
+      `\`${action}\` is not available in the \`${api}\` API. ` +
+        `Please use: ${allowedIn.map(a => `\`${a}\``).join(`, `)}.`
+    )
+  }
+
+const nodeAPIs = Object.keys(require(`../../utils/api-node-docs`))
+
+const ALLOWED_IN = `ALLOWED_IN`
+const DEPRECATED_IN = `DEPRECATED_IN`
+
+const set = (availableActionsByAPI, api, actionName, action) => {
+  availableActionsByAPI[api] = availableActionsByAPI[api] || {}
+  availableActionsByAPI[api][actionName] = action
+}
+
+const mapAvailableActionsToAPIs = restrictions => {
+  const availableActionsByAPI = {}
+
+  const actionNames = Object.keys(restrictions)
+  actionNames.forEach(actionName => {
+    const action = actions[actionName]
+
+    const allowedIn = restrictions[actionName][ALLOWED_IN]
+    allowedIn.forEach(api =>
+      set(availableActionsByAPI, api, actionName, action)
+    )
+
+    const deprecatedIn = restrictions[actionName][DEPRECATED_IN]
+    deprecatedIn.forEach(api =>
+      set(
+        availableActionsByAPI,
+        api,
+        actionName,
+        withDeprecationWarning(action, api, allowedIn)
+      )
+    )
+
+    const forbiddenIn = nodeAPIs.filter(
+      api => ![...allowedIn, ...deprecatedIn].includes(api)
+    )
+    forbiddenIn.forEach(api =>
+      set(
+        availableActionsByAPI,
+        api,
+        actionName,
+        withErrorMessage(action, api, allowedIn)
```

Comment:
```suggestion
        withErrorMessage(actionName, api, allowedIn)
```
part 2 of above comment